### PR TITLE
Started on Eval. Wrote evalInt.

### PR DESCRIPTION
**Pull request by lukasberglund**
Is this what you had in mind as eval?

### CODE REVIEW

File: src/Eval.hs [hunk 1]
```diff
@@ -0,0 +1,17 @@
+module Eval where
+
+import Grammar
+
+evalInt :: Expression -> Int
```
Comment (ayazhafiz) [inline src/Eval.hs]:
```suggestion
evalNumber :: Expression -> Int
```

Comment (lukasberglund) [reply src/Eval.hs]: But what about floats?

Comment (ayazhafiz) [reply src/Eval.hs]: i'd prefer to maintain consistency with our grammar names. we can add support for floats later on

Comment (lukasberglund) [reply src/Eval.hs]: But evalInt has to return int since Haskell is strongly typed. Are you suggesting we change the name back when we add floats?

Comment (ayazhafiz) [reply src/Eval.hs]: Let's see what happens with your change to the `Number` grammar first. If we move to a string this won't work anyway.

Comment (lukasberglund) [reply src/Eval.hs:5]: So what do you think?

Comment (ayazhafiz) [reply src/Eval.hs:5]: Since we chose to change the grammar of a number to hold a string, this won't work yet. We need to add an intermediate step to translate the number into a particular type before we can evaluate it. This will actually require some work... so let's do that first.

Basically given some expression like `1+1`, right now we can parse it to `Number(1) + Number(1)`. We need some step that can infer this expression to be `Int(1) + Int(1)`. In a later step we can add validation and errors for mismatched types as well; I think here it is enough to make inferences for the `+, -, /, *, %` functions. Then later on we can check that the parameters passed to the functions are as expected. Does that make sense?

So basically I suggest the flow right now should be

```
Parser -> Type Resolution --> Type Checking (add later) --> Eval
```

Comment (lukasberglund) [reply src/Eval.hs:5]: Makes sense. If we are doing lazy evaluation by default we only have to evaluate these expressions when it's necessary. Does this factor into what we should be doing now?

Also, what would type resolution look like?

I'm imagining some process where we add a type to each expression-branch in the AST. Is that right?

Comment (ayazhafiz) [reply src/Eval.hs:5]: Not sure, I'm actually not super familiar with how to implement lazy evaluation on the compiler side. My assumption is that it will only be for the evaluation phase, not anything before (type reso, checking, etc)

For type resolution, yes. It can just be a granular grammar with type fields.

**Comment by lukasberglund** [comment]
I ~~added a naive implementation of a typechecker~~ singlehandedly solved typechecking. PTAL!

**Comment by lukasberglund** [comment]
I just wanted to know if you thought my approach is good.

File: src/Eval.hs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+driveEval :: String -> Either Error String
+driveEval s = pure (Lexer.lex s) >>= parse >>= (showResult . head)
+
+showResult :: Expression -> Either Error String
+showResult expr = do
+  ty <- expressionType expr
+  case ty of
+    Int _ _ -> return $ show $ evalInt expr
+    Double -> return $ show $ evalDouble expr
```
Comment (ayazhafiz) [inline src/Eval.hs:17]: should probably go in the test file?

Comment (ayazhafiz) [reply src/Eval.hs:17]: I guess `showResult` could be called `eval` where we do the type checking then route to the appropriate eval fn

File: src/TypeCheck.hs [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+expressionType expr = fmap head (expressionTypes expr)
+  where
+    expressionTypes :: Expression -> Either Error [Type]
+    expressionTypes (Number s) =
+      if '.' `elem` s
+        then Right [Double]
+        else Right numTypes
+    expressionTypes (String _) = Right [Str]
+    expressionTypes (Char _) = Right [CharTy]
+    expressionTypes (Call name [a, b]) =
```
Comment (ayazhafiz) [inline src/TypeCheck.hs:28]: can you add a todo to support arbitrary # parameters

File: src/TypeCheck.hs [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+      pure operationTypes >>= filterTypes a >>= filterTypes b
+      where
+        operationTypes
+          | name `elem` ["+", "-", "*", "/"] = numTypes
+          | name == "%" = allInts
```
Comment (ayazhafiz) [inline src/TypeCheck.hs:33]: Actually maybe we should just do the typechecking for numbers and calls rn since that's what eval cares about

File: src/Eval.hs (hunk 2 above)
Comment (lukasberglund) [reply src/Eval.hs:17]: So do we want eval to always return a string?

File: src/TypeCheck.hs (hunk 4 above)
Comment (lukasberglund) [reply src/TypeCheck.hs:33]: You mean before we do eval?

File: src/Eval.hs (hunk 2 above)
Comment (ayazhafiz) [reply src/Eval.hs:17]: No I think the return type is fine I just think the name could be better.

File: src/TypeCheck.hs (hunk 4 above)
Comment (ayazhafiz) [reply src/TypeCheck.hs:33]: yes

File: hgb.cabal [hunk 5]
```diff
@@ -1,81 +0,0 @@
-cabal-version: 1.12
```
Comment (ayazhafiz) [inline hgb.cabal:1]: ?? why delete this?

File: src/Eval.hs [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+evalDouble :: Expression -> Float
+evalDouble (Number n) = read n :: Float
+evalDouble (Call "+" [x]) = evalDouble x
+evalDouble (Call "-" [x]) = -evalDouble x
+evalDouble (Call function args) =
+  case function of
+    "+" -> a + b
+    "-" -> a - b
+    "*" -> a * b
+    "/" -> a / b
```
Comment (ayazhafiz) [inline src/Eval.hs:49]: What about mod for doubles?

Comment (ayazhafiz) [reply src/Eval.hs:49]: Ah it's only ints

File: src/TypeCheck.hs [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+      case types `intersect` exprTypes of
+        [] -> wrapTypeError types expr
+        ts -> return ts
+
+functionTypes :: String -> Either Error [Type]
+functionTypes "+" = Right numTypes
+functionTypes "-" = Right numTypes
+functionTypes "*" = Right numTypes
+functionTypes "/" = Right numTypes
+functionTypes "%" = Right allInts
```
Comment (ayazhafiz) [inline src/TypeCheck.hs:46]: What if we make a registry of function names to their type signature and implementation? It will make it easier to typecheck custom functions later and have a common place for the type signature and implementation rather than splitting between typecheck and eval. Just something to think about, we can do it in another PR

File: src/TypeCheck.hs [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+    expressionTypes :: Expression -> Either Error [Type]
+    expressionTypes (Number s) =
+      if '.' `elem` s
+        then Right [Double]
+        else Right numTypes
+    expressionTypes (String _) = Right [Str]
+    expressionTypes (Char _) = Right [CharTy]
+    expressionTypes (Call name args) = do
+      funTypes <- functionTypes name
+      foldrM filterTypes funTypes args
```
Comment (ayazhafiz) [inline src/TypeCheck.hs:31]: can you add a TODO that this should (1) return the return type of the function and (2) should compare each argument with each parameter type rather than the list of all parameter types of the function

File: src/TypeCheck.hs [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+
+import Data.Foldable (foldrM)
+import Data.List (intersect)
+import Error
+import Grammar (Bits(..), Expression(..), Type(..))
+import Utils
+
+wrapTypeError :: [Type] -> Expression -> Either Error a
+wrapTypeError expectedTypes expression =
+  Left $ Error (TypeError expectedTypes expression) (Span 0 0)
```
Comment (ayazhafiz) [inline src/TypeCheck.hs:13]: TODO: fix the span?

Comment (lukasberglund) [reply src/TypeCheck.hs:13]: I can add that as a TODO! I can't actually fix the span yet since expressions aren't linked to code.

File: hgb.cabal (hunk 5 above)
Comment (lukasberglund) [reply hgb.cabal:1]: Ill add it back

File: src/TypeCheck.hs (hunk 7 above)
Comment (lukasberglund) [reply src/TypeCheck.hs:46]: That could work.

File: src/TypeCheck.hs (hunk 8 above)
Comment (lukasberglund) [reply src/TypeCheck.hs:31]: Ah true, lol I actually thought that this was was the function was gonna do. I honestly don't know what I was thinking.

**Comment by ayazhafiz** [comment]
@lukasberglund do you want a review on this?
